internal/cmd/gocdk/internal/docker/docker.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package docker provides a client for interacting with the local Docker
+// daemon. This currently shells out to the Docker CLI, but could use the HTTP
+// API directly in the future.
+package docker
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// Client is a client for a Docker daemon. The zero value is a Docker client
+// that uses the current process's environment. Methods on Client are safe to
+// call concurrently from multiple goroutines.
+type Client struct {
+	env []string
+}
+
+// New creates a new client with the given environment variables.
+// Calling docker.New(nil) is the same as new(docker.Client) and uses the
+// current process's environment.
+func New(env []string) *Client {
+	return &Client{env: env}
+}
+
+// Image stores metadata about a Docker image.
+type Image struct {
+	ID         string
+	Repository string
+	Tag        string
+	Digest     string
+	CreatedAt  time.Time
+}
+
+// ListImages list images the Docker daemon has stored locally that match filterRef.
+func (c *Client) ListImages(ctx context.Context, filterRef string) ([]Image, error) {
+	args := []string{"images", "--digests", "--format={{json .}}"}
+	if filterRef != "" {
+		args = append(args, filterRef)
+	}
+	cmd := exec.CommandContext(ctx, "docker", args...)
+	cmd.Env = c.env
+	stderr := new(bytes.Buffer)
+	cmd.Stderr = stderr
+	out, err := cmd.Output()
+	if err != nil {
+		if stderr.Len() > 0 {
+			return nil, xerrors.Errorf("docker images:\n%s", stderr)
+		}
+		return nil, xerrors.Errorf("docker images: %w", err)
```

Comment:
Do we need these prefixes on the errors? E.g., the actual error will now look something like:
`gocdk build: docker images: blah blah`. Isn't `gocdk build:` enough? If anything, it might be useful to log the exact command line.

Reply:
It is helpful for `gocdk launch` for instance, where there is a `docker tag` followed by a `docker push`, to know the operation that failed. Exact command line can be a little verbose, so this seemed like a reasonable compromise.

For this exact operation, I am seeing that `docker images` isn't as helpful as `list docker images`, so changed to that.

Reply:
OK.